### PR TITLE
Implement Database Synchronization Function for Game Playtime

### PR DESCRIPTION
**Pull request by bagusnl**
# Main Goal
Synchronize user game playtime to a database that the user can sync with other systems with privacy in mind.

Every user needs to have their own SQLite server (eg. Turso https://turso.tech/ free account should be sufficient) then supply all the necessary informations to Collapse (Database URL in https and the token). Collapse will generate the GUIDv7 as a user identifier table inside the database.

> [!CAUTION]
> Changing User ID without taking notes of your current ID will cause the data to be dangling in the database and not being used anymore.
> While you can edit the database yourself, it's a bit tedious, so please do NOT lose your user ID.

Playtime synced to the database in a few mechanism:
1. At first region load
> When region is loaded, playtime stats is first loaded from your local machine, then the system fetch and compare when the last time that region play time is updated, if the local is out of date then database value will be used and saved locally. If the database is out of date, then local value will be used and saved to the database.

2. At game first launch (force push)
3. Every 5 minutes when the game is running
4. At game close (force push)

## PR Status :
- Overall Status : In Progress
- Commits : In Progress
- Synced to base (Collapse:main) : Yes
- Build status : OK
- Crashing : Yes
- Bug found caused by PR : 0

### CODE REVIEW

**Review by gablm — CHANGES_REQUESTED** [review]
The core functionality seems to be working fine.
However, there are some problems related to error handling and file initialisation when running Collapse for the first time.

<details>
<summary>Some Front-end related suggestions</summary>

- Allow DB Url to not contain `https://` or have a different protocol from `https://`

As the database link copied from Turso is in the format `libsql://example.turso.io`, it would be nice to be able to input the url directly like that.

- Show a dialog after closing game is syncing wasn't successful.
- Show a dialog if database connection is unsuccessful at start-up.
</details>

File: CollapseLauncher/Classes/Helper/Database/DBConfig.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBConfig.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/31142aec1afb0b998dc8161578fbcc539a1ea1d6

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/1e22f9532a54a818fdcd63edbac6f4190fca339c

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:106]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/1e22f9532a54a818fdcd63edbac6f4190fca339c

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:284]: DBSync is now follow the settings https://github.com/CollapseLauncher/Collapse/pull/586/commits/2670044715ee1dbdc73df65b4ca06f8688beee76

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:341]: Fixed

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:1441]: I think it's redundant as changing the text in the textbox will invoke DbUserId.set which do have the DBConfig setter

File: CollapseLauncher/Classes/Helper/Database/DBConfig.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBConfig.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/1e22f9532a54a818fdcd63edbac6f4190fca339c

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (gablm) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:1439]: It doesn't seem to be calling the setter inside the DBConfig tho. The change is purely visual and will revert to the earlier ID when reloading the page / opening Collapse again

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (gablm) [inline CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:313]: Needs also to pass _gameSettings to the new instance.

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (gablm) [inline CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:1391]: After changing the ID, the confirmation dialog will appear when clicking any other button/toggle, the click action won't be executed in the case of the button but it will if its a toggle.

This can create a bit of user confusion.

**Review by gablm — CHANGES_REQUESTED** [review]
It's mostly working as expected but still has some small problems

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:1391]: I'm not sure on how to fix this behavior since this seems to be another WinUI shenanigans

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:1365]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/81236f83974f07df77d889713bb191660cbf3a16

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/dd77b16fd8e6ba33cabd57671b402b8e92fe9ab3

Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/dd77b16fd8e6ba33cabd57671b402b8e92fe9ab3

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml:1119]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/cfcab0cf47d0bedfe1900af4ccc5e69ee0ef5ff8

File: CollapseLauncher/XAMLs/MainApp/Pages/HomePage.xaml
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/HomePage.xaml]: Fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/cfcab0cf47d0bedfe1900af4ccc5e69ee0ef5ff8

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:313]: fixed in https://github.com/CollapseLauncher/Collapse/pull/586/commits/c8456999ca89f52a457ff56f5c587e849e33e8c6

**Comment by bagusnl** [comment]
@CollapseLauncher/core-maintainers please finish review in 12 hours if you can, planning to merge this PR on the weekend

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (shatyuka) [inline CollapseLauncher/Classes/Helper/Database/DBHandler.cs:162]: Catch some typical exceptions like "Unauthorized" to make it more user-friendly.

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (shatyuka) [inline CollapseLauncher/Classes/Helper/Database/DBHandler.cs:93]: Since the table name is hashed, is it necessary to have the user input the user id with uuid? You can add salt if you feel it's easily get reversed.

File: CollapseLauncher/XAMLs/MainApp/Pages/HomePage.xaml
Comment (shatyuka) [inline CollapseLauncher/XAMLs/MainApp/Pages/HomePage.xaml:2452]: Margin of `SyncDbPlaytimeBtn` should be `-4,0,-4,-4`, and `ResetPlaytimeButton` margin `-4,0`

**Review by shatyuka — CHANGES_REQUESTED** [review]
Made a quick review.

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (Cryotechnic) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:1391]: Resolving since it seems to be an issue with WinUI's implementation.

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (Cryotechnic) [inline CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:1490]: Whitespace

File: Hi3Helper.Core/Lang/en_US.json
Comment (Cryotechnic) [inline Hi3Helper.Core/Lang/en_US.json:647]: `Examples: https://db-collapse.turso.io` so that it represents the project

File: Hi3Helper.Core/Lang/en_US.json
Comment (Cryotechnic) [inline Hi3Helper.Core/Lang/en_US.json:652]: `could not`->`cannot`

File: Hi3Helper.Core/Lang/en_US.json
Comment (Cryotechnic) [inline Hi3Helper.Core/Lang/en_US.json:653]: See previous

File: Hi3Helper.Core/Lang/en_US.json
Comment (Cryotechnic) [inline Hi3Helper.Core/Lang/en_US.json:967]: `Are you sure you wish to change your user ID?`

File: Hi3Helper.Core/Lang/en_US.json
Comment (Cryotechnic) [inline Hi3Helper.Core/Lang/en_US.json:968]: `Changing the current user ID will cause the associated data to be lost if you lose it.`

**Review by Cryotechnic — CHANGES_REQUESTED** [review]
Overall looks ok, I share some concerns with shatyuka as well as some other things.

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (gablm) [inline CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:313]: Deserialization doesn't seem to be working, always returning an instance with default values.

**Review by gablm — CHANGES_REQUESTED** [review]
Functionality wise everything seems to be working apart from statistic syncing.
I agree with cryo that could be good to show more errors in the UI when they occur.

Related to changing the url/token/uid in the settings, I believe it would be better to have a "save & validate" button instead of instantly changing the fields. 
As it is, it is possible to insert a incorrect url and/or token and no error would ever be displayed if the user never clicks the "Validate Settings" button.
Another possible solution is displaying a dialog when first connection fails.

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:93]: I was thinking it on how easy it is to generate the GUID on hand and parse it to see if its a valid one without compromising the security of the database if the table names were dumped (hashed guid can't directly related to a user, unlike a direct user-inputted user ID). The reverse-ability in the comment is purely a joke.

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:166]: As the connection is really small, I don't think it's necessary to put detection for metered in this system

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:161]: This error will never be displayed as public because I want to set the database system to be completely in the background and non-invasive to the UX. I tried to do push in-app notifications but it seems currently the caller for it doesn't work outside HomePage methods (maybe i am dum, probably, most likely)

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:282]: When `redirectThrow` is set to true, any exception will be thrown back to the method caller. This is (currently) only used in the settings validation as I need a way to get what errors to be displayed to the user.

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:221]: Copied my response from other thread 
> When `redirectThrow` is set to true, any exception will be thrown back to the method caller. This is (currently) only used in the settings validation as I need a way to get what errors to be displayed to the user. The only difference is just it has `throw` instead of `return`

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:313]: i can replicate it, but im not sure on why this is the case as the return bytes of the db query is perfect (in my eyes)
Example of json query result
```
7B 22 44 61 69 6C 79 50 6C 61 79 74 69 6D 65 22 3A 22 30 30 3A 34 30 3A 30 30 22 2C 22 57 65 65 6B 6C 79 50 6C 61 79 74 69 6D 65 22 3A 22 32 33 3A 30 30 3A 30 30 22 2C 22 4D 6F 6E 74 68 6C 79 50 6C 61 79 74 69 6D 65 22 3A 22 32 36 3A 30 30 3A 30 30 22 2C 22 4C 61 73 74 53 65 73 73 69 6F 6E 22 3A 22 30 30 3A 30 30 3A 30 30 22 2C 22 43 6F 6E 74 72 6F 6C 44 61 74 65 22 3A 22 32 30 32 34 2D 31 30 2D 31 33 54 30 30 3A 30 30 3A 30 30 2B 30 37 3A 30 30 22 7D
```
image


I'm really not sure how this happened now since it was fine when me and @neon-nyan tested it a few weeks ago (i think we tested it)

File: CollapseLauncher/XAMLs/MainApp/Pages/HomePage.xaml
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/HomePage.xaml]: Fixed

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs]: Fixed

File: Hi3Helper.Core/Lang/en_US.json
Comment (bagusnl) [reply Hi3Helper.Core/Lang/en_US.json]: Fixed

File: Hi3Helper.Core/Lang/en_US.json
Comment (bagusnl) [reply Hi3Helper.Core/Lang/en_US.json]: Fixed

File: Hi3Helper.Core/Lang/en_US.json
Comment (bagusnl) [reply Hi3Helper.Core/Lang/en_US.json]: Fixed

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (gablm) [reply CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:312]: this was actually my fault, it was already broken in #584
should be working now

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (shatyuka) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: > I was thinking it on how easy it is to generate the GUID on hand and parse it to see if its a valid one without compromising the security of the database if the table names were dumped (hashed guid can't directly related to a user, unlike a direct user-inputted user ID). The reverse-ability in the comment is purely a joke.

Since the user id is hashed, it's not recognizable for hackers. It's also more userfriendly cuz we don't need to sync the GUID in another way.

Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: Got it, I'll try to move to random word generator and see how it does 

Ref https://stackoverflow.com/questions/18110243/random-word-generator-2

File: CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs
Comment (Cryotechnic) [reply CollapseLauncher/Classes/GameManagement/GamePlaytime/RegistryClass/CollapsePlaytime.cs:165]: We can leave it as-is for now and adjust if needed then.

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (shatyuka) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: I think we can just let user input anything they want, and we hash it with salt like "collaspe-". Also sqlite does not have permission system, so one should use his own server. Do we really need a user id?

Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: UserID still needed for the instance where a user wants to have a shared database server but doesn't share the data

Comment (shatyuka) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: So now the security is based on the RANDOM UUID to get rid of being overrided by others with same user id. :(

Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: I'm not sure what you meant, my brain is melting due to heat today. 
But we have a few choices for this one
1. Keep the logic as is, where the UserID uses UUIDv7 that is hashed. This is simpler logically and more secure, but not very user friendly as UUIDv7 is pretty long.
2. Move from UUIDv7 to random words (still editable by users) that is hashed. Should be more user friendly but maybe less secure since user might just put a very simple UserID.

Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: I'm not sure what you meant, my brain is melting due to heat today. 
But we have a few choices for this one
1. Keep the logic as is, where the UserID uses UUIDv7 that is hashed. This is simpler logically and more secure, but not very user friendly as UUIDv7 is pretty long.
2. Move from UUIDv7 to random words (still editable by users) that is hashed. Should be more user friendly but maybe less secure since user might just put a very simple UserID. Tbh this still sufficient since its the user choice to share their database so, eh.

Comment (shatyuka) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:98]: generate UUID, but editable by users and not limited to UUID format.

File: Hi3Helper.Core/Lang/en_US.json
Comment (bagusnl) [reply Hi3Helper.Core/Lang/en_US.json]: Fixed

File: Hi3Helper.Core/Lang/en_US.json
Comment (bagusnl) [reply Hi3Helper.Core/Lang/en_US.json]: Fixed

File: CollapseLauncher/Classes/Helper/Database/DBHandler.cs
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Database/DBHandler.cs:100]: Adjusted in https://github.com/CollapseLauncher/Collapse/pull/586/commits/2dede52ab45a50354713637ba728e9e62f77c8df

**Comment by bagusnl** [comment]
After discussing with neon in chat, I'm merging this PR. Should there be any concern of the code, hit me up in this PR comment or directly to me

**Pull request merged**
